src/Extension/Chrome/Browser.php

```diff
@@ -0,0 +1,150 @@
+    /** @var Deferred[] */
+    private $registry = [];
+
+    /** @var Session[] */
+    private $sessions = [];
+
+    /** @var LoggerInterface */
+    private $logger;
+
+    public function __construct(string $endpoint, LoggerInterface $logger)
+    {
+        $this->endpoint = $endpoint;
+        $this->logger = $logger;
+    }
+
+    public function connect()
+    {
+        return \Amp\call(function () {
+            try {
+                $this->connection = yield connect($this->endpoint);
+                $this->loop();
+
+                return true;
+            } catch (\Throwable $exception) {
+                return false;
+            }
+        });
+    }
+
+    protected function loop()
+    {
+        return \Amp\call(function () {
+            $error = false;
+
+            while ($this->connection !== null && $this->connection->valid()) {
+                /** @var Message $message */
+                $message = $this->connection->current();
+                $data = yield $message->read();
+                $this->logger->info('Receive message ' . $data);
+                $decoded = @json_decode($data, true);
+
+                if (!$decoded) {
+                    $this->logger->error('Cannot decode message ' . $data);
+                }
+
+                if ($decoded !== false && isset($decoded['id'])) {
+                    $messageId = $decoded['id'] ?? null;
+
+                    if (array_key_exists($messageId, $this->registry)) {
+                        $deferred = $this->registry[$messageId];
+                        unset($this->registry[$messageId]);
+
+                        $deferred->resolve($decoded['result']);
+                    }
+                }
+
+                if (array_key_exists('method', $decoded)) {
+                    $method = $decoded['method'];
+                    $params = $decoded['params'];
+
+                    $this->emit($method, $params);
+                }
+
+                $this->connection->next();
+            }
+
+            return $error;
+        });
+    }
+
+    public function close()
+    {
+        $this->connection->close();
+    }
+
+    protected function isConnected()
+    {
+        return $this->connection !== null;
+    }
+
+    public function send(string $method, array $params = []): Promise
+    {
+        $messageId = $this->nextMessageId;
+        $deferred = new Deferred();
+        $this->registry[$messageId] = $deferred;
+
+        \Amp\call(function () use($method, $params, $messageId) {
+            // Force {} for encoding
+            if (empty($params)) {
+                $params = new \stdClass();
+            }
+
+            $message = json_encode([
+                'id' => $messageId,
+                'method' => $method,
+                'params' => $params
+            ]);
+
+            $this->nextMessageId++;
+
+            $this->logger->info('Send message ' . $message);
+
+            yield $this->connection->send($message);
+        });
```

Comment:
Any exception inside this `Amp\call` will go unnoticed. It's better to use `Amp\asyncCall` to have them thrown into the loop or better just handle them.